Bloom.md

```diff
@@ -0,0 +1,248 @@
+
+Pros
+* This is the simplest option in terms of effort - assuming the user is fine with recreating bloom objects and adding items on the Valkey server (with ValkeyBloom).
+
+Cons
+* The user will need to re-create the bloom objects (using BF.RESERVE/BF.INSERT) and populate these objects **AFTER** switching to Valkey (with ValkeyBloom).
+
+2. Users can generate an AOF file from a server (that has the ReBloom module loaded) and with an on-going bloom workload that creates bloom filters & inserts items into them. Next, this can be re-played on a Valkey Server (with ValkeyBloom loaded). Then, the user can move their existing bloom workload to the Valkey server (with ValkeyBloom loaded).
+
+Pros
+* Bloom objects can be created on the user's existing system (with Rebloom) and the user can also populate the objects by adding items (BF.ADD/BF.MADD) **BEFORE** switching to Valkey (with ValkeyBloom).
+
+Cons
+* The user will need to manage AOF file generation on the server (with ReBloom) and ensure that the contents do not get re-written (e.g. as a result of BGREWRITEAOF) into a RDB file. This is because the RDB file with bloom object data generated by ReBloom is not compatible with ValkeyBloom. The main drawback here is that the AOF can exceed a size limit and get truncated / re-written.
+
+### Memory Management
+
+The bloom data type also supports memory management related callbacks:
+
+* free: Frees the bloom filter object when it is deleted, expired or evicted.
+* defrag: Supports active defrag for Bloom filter objects
+* mem_usage: Reports bytes used by a Bloom object and is used by the MEMORY USAGE command
+* copy: Supports a deep copy of bloom filter objects and is used by the COPY command
+* free_effort: Determine whether the bloom object's memory needs to be lazy reclaimed or synchronously freed. We return the number of filters in the bloom object as the free effort and this is similar to how the core handles free_effort for aggregated objects.
+
+
+### Replication
+
+Every Bloom Filter based write command will be replicated to replica nodes.
+
+### Keyspace Event Notification
+
+Every bloom filter based write command will be made to publish a keyspace event after the data is mutated.
+* Event type: VALKEYMODULE_NOTIFY_GENERIC
+* Event name: command name in lowercase. e.g., BF.ADD command publishes event "bloom.add".
+
+Users can subscribe to the bloom events via the standard keyspace event pub/sub. For example,
+
+```text
+1. enable keyspace event notifications:
+    valkey-cli config set notify-keyspace-events KEA
+2. suscribe to keyspace & keyevent event channels:
+    valkey-cli psubscribe '__key*__:*'
+```
+
+### Scalable filters / Non Scaling filters
+
+Bloom Filters can either be configured as scalable (default) or non-scalable through specification with the BF.RESERVE or BF.INSERT command.
+
+When non-scaling filters reach their capacity, if a user tries to add items to the bloom object, an error is returned.
```

Comment:
Why is an error returned? Wouldn't it also be valid to just return an OK back? If someone is aware of the tradeoffs associated with a bloom filter, that is can return false negatives, it seems OK that we could saturate it.

Reply:
> Why is an error returned?

This is the behavior built into ReBloom based on which this error logic was implemented. 

> Wouldn't it also be valid to just return an OK back? If someone is aware of the tradeoffs associated with a bloom filter, that is can return false negatives, it seems OK that we could saturate it.

Originally, I felt the same. But this can lead to the user observing high error_rate. It might be better for the user to be notified of this happening.

If we think that saturated bloom filters (with low occurrences of "new items" being added) are a use case, we can consider this being a configurable behavior

Reply:
Let's add that as part of the spec then. I don't necessarily want to copy RedisBloom if they made bad decisions.